chemicalx/models/gcnbmp.py

```diff
@@ -1,14 +1,255 @@
+        proj_result = F.relu(self.proj(concat_inputs))
+        proj_gate = F.sigmoid(self.transform(concat_inputs))
+        gated = (proj_gate * proj_result) + ((1 - proj_gate) * input)
+        return gated
+
+
+class AttentionPooling(nn.Module):
+    def __init__(self, molecule_channels, hidden_channels):
+        """Instantiate the Attention pooling layer
+        :param molecules_channels (int): input node features (layer 0 of the backbone).
+        :param hidden_channels (int): final node representation (layer L of the backbone).
+        """
+        super(AttentionPooling, self).__init__()
+        total_features_channels = molecule_channels + hidden_channels
+        self.lin = nn.Linear(
+            total_features_channels, hidden_channels
+        )  # weights here must be shared across all nodes according to the paper
+        self.last_rep = nn.Linear(hidden_channels, hidden_channels)
+
+    def forward(self, input_rep, final_rep, graph_index):
+        """
+        Compute an attention-based readout using the input and output layers of the 
+        RGCN encoder for one molecule.
+
+        Parameters:
+            input_rep (Tensor): input nodes representations
+            final_rep (Tensor): final nodes representations
+            graph_index (Tensor): node to graph readout index
+
+        Returns:
+            g (Tensor): graph-level representation
+        """
+        att = torch.sigmoid(self.lin(torch.cat((input_rep, final_rep), 1)))
+        g = att.mul(self.last_rep(final_rep))
+        g = scatter_add(g, graph_index, dim=0)
+        return g
+
+
+class GCNBMPEncoder(nn.Module, core.Configurable):
+    def __init__(
+        self,
+        input_dim,
+        hidden_dims,
+        num_relation,
+        edge_input_dim=None,
+        short_cut=False,
+        batch_norm=False,
+        activation="sigmoid",
+        concat_hidden=False,
+    ):
+        """Instantiate the GCN-BMP encoder.
+        :param input_dim (int): input dimension.
+        :param hidden_dims (list of int): hidden dimensions.
+        :param output_dim (int): output dimension.
+        :param num_relation (int): number of relations.
+        :param edge_input_dim (int, optional): dimension of edge features.
+        :param batch_norm (bool, optional): apply batch normalization on nodes or not.
+        :param activation (str or function, optional): activation function.
+        """
+        super(GCNBMPEncoder, self).__init__()
+
+        if not isinstance(hidden_dims, Sequence):
+            hidden_dims = [hidden_dims]
+        self.input_dim = input_dim
+        self.output_dim = hidden_dims[-1] * (len(hidden_dims) if concat_hidden else 1)
+        self.dims = [input_dim] + list(hidden_dims)
+        self.num_relation = num_relation
+        self.short_cut = short_cut
+        self.concat_hidden = concat_hidden
+
+        self.layers = nn.ModuleList()
+        for i in range(len(self.dims) - 1):
```

Comment:
use `from more_itertools import pairwise` and then do

```suggestion
        for left_dim, right_dim in pairwise(self.dims):
```